app/src/main/kotlin/com/wire/android/ui/authentication/devices/remove/RemoveDeviceViewModel.kt

```diff
@@ -78,16 +71,64 @@ class RemoveDeviceViewModel @Inject constructor(
         updateStateIfDialogVisible { it.copy(error = RemoveDeviceError.None) }
     }
 
-    fun onItemClicked(device: Device) {
-        updateStateIfSuccess { it.copy(removeDeviceDialogState = RemoveDeviceDialogState.Visible(device = device)) }
+    fun onItemClicked(client: Client) {
+        tryToDeleteOrShowPasswordDialog(client)
+    }
+
+    private fun tryToDeleteOrShowPasswordDialog(client: Client) {
+        // try delete with no password (will success only for SSO accounts)
+        viewModelScope.launch(Dispatchers.Main) {
+            // try delete with no password (will success only for SSO accounts
+
+            when (val deleteResult = deleteClientUseCase(DeleteClientParam(null, client.id))) {
+
+                DeleteClientResult.Success -> registerClientUseCase(
+                    RegisterClientUseCase.RegisterClientParam(null, null)
+                ).also { result ->
+                    when (result) {
+                        is RegisterClientResult.Failure.PasswordAuthRequired -> updateStateIfSuccess {
+                            it.copy(
+                                removeDeviceDialogState = RemoveDeviceDialogState.Visible(
+                                    client = client
+                                )
+                            )
+                        }
+                        is RegisterClientResult.Failure.Generic -> state = RemoveDeviceState.Error(result.genericFailure)
+                        RegisterClientResult.Failure.InvalidCredentials -> {
+                            // server will never response with InvalidCredentials in case of trying with no password
+                        }
+                        RegisterClientResult.Failure.TooManyClients -> {}
+                        is RegisterClientResult.Success -> {
+                            registerPushToken(result.client.id)
+                            navigateToConvScreen()
+                        }
+                    }
+                }
+                is DeleteClientResult.Failure.Generic -> state = RemoveDeviceState.Error(deleteResult.genericFailure)
+                DeleteClientResult.Failure.InvalidCredentials -> updateStateIfSuccess {
+                    it.copy(
+                        removeDeviceDialogState = RemoveDeviceDialogState.Visible(
+                            client = client
+                        )
+                    )
+                }
```

Comment:
```
updateStateIfSuccess {
                    it.copy(
                        removeDeviceDialogState = RemoveDeviceDialogState.Visible(
                            client = client
                        )
                    )
                }
```

I know it's not a lot of lines of code but it's used several times and my pedantic inner voice tells me "maybe it's worth extracting" 😄